asustor.c

```diff
@@ -244,45 +236,101 @@ static struct gpiod_lookup_table asustor_600_gpio_keys_lookup = {
 };
 // clang-format on
 
+struct pci_device_match {
+	// match PCI devices with the given vendorID and productID (to help identify ASUSTOR systems)
+	// you can get them from `lspci -nn`, for example in
+	// "00:08.0 System peripheral [0880]: Intel Corporation Device [8086:4e11]"
+	// 0x8086 is the vendorID and 0x4e11 is the deviceID
+	uint16_t vendorID;
+	uint16_t deviceID;
+
+	int16_t min_count; // how often that device should exist at least
+	int16_t max_count; // how often that device should exist at most
+};
+
 // ASUSTOR Platform.
 struct asustor_driver_data {
+	const char *name; // used for force_device and for some log messages
+
+	struct pci_device_match pci_matches[4];
+
 	struct gpiod_lookup_table *leds;
 	struct gpiod_lookup_table *keys;
 };
 
+#define VALID_OVERRIDE_NAMES "AS6xx, AS61xx, AS66xx, AS67xx, FS67xx"
+
+// NOTE: if you add another device here, update VALID_OVERRIDE_NAMES accordingly!
+
 static struct asustor_driver_data asustor_fs6700_driver_data = {
+	.name = "FS67xx",
+	// FS67xx needs to match PCI devices because it has the same DMI data as *A*S67xx
+	.pci_matches = {
+		// PCI bridge: ASMedia Technology Inc. ASM2806 4-Port PCIe x2 Gen3 Packet Switch (rev 01)
+		// FS6712X seems to have 15 of these, no idea about FS6706T (so I keep min_count at 1)
+		// currently the upper limit doesn't matter so I just use 10000
+		{ 0x1b21, 0x2806, 1, 10000 },
```

Comment:
These are just the numbers for this specific device.
If only 1, 16 and 0, 0 were relevant numbers, I could use a bool..

For the next PCI device we match it could be that one series of NAS uses 1-3 of them and the other 5-10

Reply:
I hear you. I wasn't suggesting these are the only options for values and agree that boolean would be too limited. I'm not married to any particular solution though, just gave a suggestion as I thought clarity could be improved (especially with the max number). We could for instance let max be 0 which I think would say more than 10,000 about its purpose.

Reply:
0 is needed for "none at all" (vs 0 to MAX for "don't care how many"), but maybe -1 would be an option.. OTOH that introduces a special case in the code.
As its saved in a int16_t, INT16_MAX could be used?

Reply:
Sure, why not, let's go with `INT16_MAX` 😄

Reply:
ok :)

Reply:
unfortunately it turns out that apparently <stdint.h> doesn't exist for the kernel

Reply:
I introduced DEVICE_COUNT_MAX for this instead